File: backend/app/services/priority.py
"""Priority scoring engine combining demand, population impact, and infrastructure gap."""

from typing import Dict, Any


def calculate_priority_score(
    demand_score: float,
    population_impact_score: float,
    infra_gap_score: float,
    investment_score: float = 50.0,
) -> Dict[str, Any]:
    """Calculates multi-dimensional priority score from 0 to 100."""
    # Weights: Demand (35%), Population Impact (30%), Infrastructure Deficit (25%), Investment Deficit (10%)
    investment_deficit = max(0.0, 100.0 - investment_score)

    composite = (
        (demand_score * 0.35)
        + (population_impact_score * 0.30)
        + (infra_gap_score * 0.25)
        + (investment_deficit * 0.10)
    )

    if composite >= 80.0:
        priority_label = "critical"
    elif composite >= 65.0:
        priority_label = "high"
    elif composite >= 45.0:
        priority_label = "medium"
    else:
        priority_label = "low"

    return {
        "priorityScore": round(composite, 1),
        "priority": priority_label,
    }

File: backend/functions/src/services/priority.ts
import { Priority } from '../models/types';

export function calculatePriorityScore(
  demandScore: number,
  populationImpactScore: number,
  infraGapScore: number,
  investmentScore: number = 50.0
): { priorityScore: number; priority: Priority } {
  const investmentDeficit = Math.max(0.0, 100.0 - investmentScore);

  // Weights: Demand (35%), Population Impact (30%), Infrastructure Deficit (25%), Investment Deficit (10%)
  const composite =
    demandScore * 0.35 +
    populationImpactScore * 0.3 +
    infraGapScore * 0.25 +
    investmentDeficit * 0.1;

  let priority: Priority = 'low';
  if (composite >= 80.0) {
    priority = 'critical';
  } else if (composite >= 65.0) {
    priority = 'high';
  } else if (composite >= 45.0) {
    priority = 'medium';
  }

  return {
    priorityScore: Math.round(composite * 10) / 10,
    priority,
  };
}
